module/bdev/ocf/ctx.c

```diff
@@ -533,6 +539,195 @@ vbdev_ocf_ctx_log_printf(ocf_logger_t logger, ocf_logger_lvl_t lvl,
 	return 0;
 }
 
+struct ocf_cache_persistent_meta_segment_descriptor {
+	int id;
+
+	bool valid;
+	size_t size;
+	size_t offset;
+};
+
+#define MAX_SEGMENTS 20
+struct shm_superblock {
+	struct ocf_cache_persistent_meta_segment_descriptor segments[MAX_SEGMENTS];
+};
+
+
+static ocf_persistent_meta_zone_t
+vbdev_ocf_persistent_meta_init(ocf_cache_t cache, size_t size, bool *load)
+{
+	struct vbdev_ocf_cache_ctx *ctx = ocf_cache_get_priv(cache);
+	struct ocf_persistent_meta_zone *pmeta;
+	struct stat stat_buf;
+	void *shm;
+	struct shm_superblock *shm_sb;
+	int zone;
+
+	size += sizeof(struct shm_superblock);
+
+	if (!ctx->create && ctx->force)
+		return NULL;
+
+	for (zone = 0; zone < MAX_PERSISTENT_ZONES; zone++) {
+		if (!ctx->persistent_meta[zone].fd)
+			break;
+	}
+
+	if (zone >= MAX_PERSISTENT_ZONES)
+		return NULL;
+
+
+	pmeta = &ctx->persistent_meta[zone];
+	*load = false;
+
+	snprintf(pmeta->name, NAME_MAX, "/ocf.%s.%u", ctx->cache_name, zone);
+
+	pmeta->fd = shm_open(pmeta->name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+	if (pmeta->fd < 0) {
+		SPDK_ERRLOG("Can't open SHM %s\n", pmeta->name);
+		return NULL;
+	}
+
+	if (ctx->force) {
+		SPDK_NOTICELOG("Using force, truncating SHM %s\n",
+				pmeta->name);
+		if (ftruncate(pmeta->fd, 0) < 0) {
+			SPDK_ERRLOG("Failed to truncate\n");
+			goto err_stat;
+		}
+
+	} else if (fstat(pmeta->fd, &stat_buf) < 0) {
+		SPDK_ERRLOG("Can't stat SHM %s\n", pmeta->name);
+		goto err_stat;
+	}
+
+	if (!ctx->force && stat_buf.st_size != 0) {
+		if (ctx->create) {
+			SPDK_ERRLOG("SHM %s found, but no force specified!\n",
+					pmeta->name);
+			goto err_stat;
+		}
+
+		*load = true;
+		SPDK_NOTICELOG("Loading from SHM\n");
+		if (stat_buf.st_size < size) {
+			if (ftruncate(pmeta->fd, size) < 0) {
+				SPDK_ERRLOG("Failed to extend SHM\n");
+				goto err_stat;
+			}
+			pmeta->size = size;
+		} else if (stat_buf.st_size == size) {
+			pmeta->size = size;
+		} else {
+			SPDK_NOTICELOG("Refusing to shrink SHM\n");
+			pmeta->size = stat_buf.st_size;
+		}
+	} else {
+		SPDK_NOTICELOG("Set SHM size\n");
+		if (ftruncate(pmeta->fd, size) < 0) {
+			SPDK_ERRLOG("Failed to truncate\n");
+			goto err_stat;
+		}
+
+		pmeta->size = size;
+	}
+
+	shm = mmap(NULL, pmeta->size, PROT_READ | PROT_WRITE, MAP_SHARED, pmeta->fd, 0);
+	if (shm == MAP_FAILED) {
+		SPDK_ERRLOG("Failed to map shm\n");
+		goto err_stat;
+	}
+
+	pmeta->data = shm;
+	shm_sb = shm;
+
+	if (mlock(shm, pmeta->size)) {
+		SPDK_ERRLOG("Failed to mlock\n");
+
+		goto err_mlock;
```

Comment:
If the locking failed, shouldn't this skip executing munlock?

Reply:
fixed